extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -584,4 +815,79 @@ public TopicPartition apply(Integer n)
         )
     );
   }
+
+  private Set<Integer> assignPartitionsAndSeekToStart(KafkaConsumer consumer, String topic)
+  {
+    // Initialize consumer assignment.
+    final Set<Integer> assignment = Sets.newHashSet();
+    for (Map.Entry<Integer, Long> entry : nextOffsets.entrySet()) {
+      final long endOffset = endOffsets.get(entry.getKey());
+      if (entry.getValue() < endOffset) {
+        assignment.add(entry.getKey());
+      } else if (entry.getValue() == endOffset) {
+        log.info("Finished reading partition[%d].", entry.getKey());
+      } else {
+        throw new ISE(
+            "WTF?! Cannot start from offset[%,d] > endOffset[%,d]",
+            entry.getValue(),
+            endOffset
+        );
+      }
+    }
+
+    assignPartitions(consumer, topic, assignment);
+
+    // Seek to starting offsets.
+    for (final int partition : assignment) {
+      final long offset = nextOffsets.get(partition);
+      log.info("Seeking partition[%d] to offset[%,d].", partition, offset);
+      consumer.seek(new TopicPartition(topic, partition), offset);
+    }
+
+    return assignment;
+  }
+
+  /**
+   * Checks if the pauseRequested flag was set and if so blocks:
+   * a) if pauseMillis == 0, until pauseRequested is cleared and shouldResume is signalled
+   * b) if pauseMillis != 0, until pauseMillis elapses -or- pauseRequested is cleared and shouldResume is signalled
+   * <p>
+   * Signals shouldPause so that callers can be notified when the pause command has been accepted.
+   * <p>
+   * Additionally, pauses if all partitions assignments have been read and getPauseAfterRead flag is set.
+   *
+   * @return true if a pause request was handled, false otherwise
+   */
+  private boolean possiblyPause(Set<Integer> assignment) throws InterruptedException
+  {
+    pauseLock.lockInterruptibly();
+    try {
+      if (ioConfig.getPauseAfterRead() && assignment.isEmpty()) {
+        pauseMillis = 0;
+        pauseRequested = true;
+      }
+
+      if (pauseRequested) {
+        log.info(pauseMillis == 0 ? "Pausing ingestion until resumed" : "Pausing ingestion for %d ms", pauseMillis);
+        paused = true;
+        shouldPause.signalAll();
+        while (pauseRequested) {
+          if (pauseMillis == 0) {
+            shouldResume.await();
+          } else if (!shouldResume.await(pauseMillis, TimeUnit.MILLISECONDS)) {
```

Comment:
`await` can return early, so you should check for that and await again if needed. `awaitNanos` makes this a bit easier as it returns the number of nanos still left to wait.